git/operations.go

```diff
@@ -329,20 +329,32 @@ func traceGitCommand(args []string, config gitCmdConfig, stdOutAndStdErr string)
 }
 
 type threadSafeBuffer struct {
-	bytes.Buffer
-	sync.Mutex
+	buf bytes.Buffer
+	mu  sync.Mutex
 }
 
 func (b *threadSafeBuffer) Write(p []byte) (n int, err error) {
-	b.Lock()
-	defer b.Unlock()
-	return b.Buffer.Write(p)
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	return b.buf.Write(p)
 }
 
-func (b *threadSafeBuffer) ReadFrom(r io.Reader) (n int64, err error) {
-	b.Lock()
-	defer b.Unlock()
-	return b.Buffer.ReadFrom(r)
+func (b *threadSafeBuffer) Read(p []byte) (n int, err error) {
+	b.mu.Lock()
```

Comment:
Could this be an `RLock()`?

Reply:
We could use a RWMutex, yes. For this particular situation, I don't think it would make a difference, because the concurrency is on the `Write`s (stderr and stdout writing to the same buffer), and `Read` and `Bytes` and `String` are not used until those are finished.

(Why was there a race between `Write` and `ReadFrom`, but not concurrency of `Write` and `Read`? Because `ReadFrom` is used when copying data to the buffer, i.e., it's a write operation)